Generalize `net::udp_channel` into `net::datagram_channel`

havaker opened the pull request:
This is a successor of: https://github.com/scylladb/seastar/pull/1804

@avikivity written in the previous PR:
> We could rename udp_channel to datagram_channel and deprecate the old name.

At first I wanted to do a basic rename, but forcing unbound&bound channel creation into the `make_udp_channel` signature turned out to be messy. 

This series:
- does a rename of udp related things
- introduces a new datagram channel creation API
- adapts `posix_udp_channel` to protocol stacks other than UDP over IP
- deprecates `udp_channel` creation methods

Fixes: #1894

avikivity commented on src/net/posix-stack.cc:
```diff
@@ -825,17 +847,23 @@ future<> posix_datagram_channel::send(const socket_address& dst, packet p) {
+        return make_unbound_datagram_channel(AF_INET);
+    }
 }
 
```
You could put this patch first, then the previous patch doesn't have the not-so-nice throw unimplemented.

havaker replied:
I agree that the throw is not so nice:/ I would still prefer to keep this order because introducing the interface first makes it easier to understand the motivation behind implementation.

avikivity commented on src/net/posix-stack.cc:
```diff
@@ -801,7 +813,7 @@ class posix_datagram_channel : public datagram_channel_impl {
     /// communicate with addresses that belong to the family of \param local.
     posix_datagram_channel(socket_address local)
-        : _closed(false) {
+        : _recv(is_inet(local.family())), _closed(false) {
```
That's a 128 byte copy. I guess it's not terrible.

avikivity commented:
Looks good.

dawmd commented on src/net/posix-stack.cc:
```diff
@@ -897,7 +909,7 @@ future<datagram>
     _recv.prepare();
     return _fd.recvmsg(&_recv._hdr).then([this] (size_t size) {
-        socket_address dst;
+        socket_address dst = _address; // Default for when no pktinfo is available (e.g. AF_UNIX).
```
Assuming the compiler doesn't optimize it (and it seems it shouldn't), it could be replaced by an IIFE to avoid copying in the loop

havaker replied:
> avoid copying in the loop

Each assignment to the `dst` in the loop is followed by `break`, so I don't really see what did you mean.

dawmd replied:
> > avoid copying in the loop
> 
> Each assignment to the `dst` in the loop is followed by `break`, so I don't really see what did you mean.

The assignment is what I meant by copying. In that version of the code, the default constructor of `socket_address` was first invoked and then followed by the copy assignment operator. It could be reduced to just one call to the right constructor by employing some kind of IFFE, e.g.
```cpp
socket_address dst = [&] {
    for (auto* cmsg = CMSG_FIRSTHDR(&_recv._hdr); cmsg != nullptr; cmsg = CMSG_NXTHDR(&_recv._hdr, cmsg)) {
        if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_PKTINFO) {
            return ipv4_addr(copy_reinterpret_cast<in_pktinfo>(CMSG_DATA(cmsg)).ipi_addr, _address.port());
        } else if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO) {
            return ipv6_addr(copy_reinterpret_cast<in6_pktinfo>(CMSG_DATA(cmsg)).ipi6_addr, _address.port());
        }
    }
    return socket_address{};
} ();
```
Here the compiler should leverage copy elision by constructing the object in place. I'm not very knowledgeable in the specifics of compilers, so maybe it's not that easy, but that was the idea I had in mind.

avikivity replied:
It's okay, copying 128 bytes is just 16 instructions (or 8 if avx is enabled). It's not material compared to the rest of the work needed to send udp (which involves a syscall as we don't have an io_uring path for it).

havaker commented:
v2:
- changed the signature of  `net::datagram_channel make_bound_datagram_channel(socket_address)` -> `net::datagram_channel make_bound_datagram_channel(const socket_address&)`
- adressed [the comment](https://github.com/scylladb/seastar/pull/1895#discussion_r1372156123) about unnecessary copying: wrapped the type of the mentioned variable into `std::optional`
- rebased to the newest master

avikivity commented on src/net/posix-stack.cc:
```diff
@@ -918,7 +930,7 @@ posix_datagram_channel::receive() {
         }
         return make_ready_future<datagram>(datagram(std::make_unique<posix_datagram>(
-            _recv._src_addr, dst, packet(fragment{_recv._buffer, size}, make_deleter([buf = _recv._buffer] { delete[] buf; })))));
+            _recv._src_addr, dst ? *dst : _address, packet(fragment{_recv._buffer, size}, make_deleter([buf = _recv._buffer] { delete[] buf; })))));
```
I actually like the previous version with unconditional copying better (small copy = faster than conditional). But I don't think it matters, I'm sure it's impossible to measure a difference.
